crates/subspace-wasm-tools/src/lib.rs

```diff
@@ -1,15 +1,38 @@
-use std::{env, fs};
+use std::{env, fs, path::PathBuf};
+
+fn get_relative_target(dir: PathBuf) -> Option<PathBuf> {
+    if dir.join("Cargo.lock").is_file() && dir.join("target").is_dir() {
+        return Some(dir.join("target"));
+    }
+    dir.parent()
+        .map(ToOwned::to_owned)
+        .and_then(get_relative_target)
+}
 
 /// Creates a `target_file_name` in `OUT_DIR` that will contain constant `bundle_const_name` with
 /// runtime of `runtime_crate_name` stored in it.
 ///
 /// Must be called before Substrate's WASM builder.
 pub fn create_runtime_bundle_inclusion_file(
-    target_dir: &str,
     runtime_crate_name: &str,
     bundle_const_name: &str,
     target_file_name: &str,
 ) {
+    let target_dir = if let Ok(target_dir) =
+        env::var("CARGO_TARGET_DIR").or_else(|_| env::var("WASM_TARGET_DIRECTORY"))
+    {
+        // Propagate target directory to wasm build
+        target_dir.into()
+    } else {
+        let out_dir = env::var("OUT_DIR").expect("Always set by cargo");
+        get_relative_target(out_dir.into())
+            .map(|target_dir| get_relative_target(target_dir.clone()).unwrap_or(target_dir))
+            .expect("Out dir is always inside the target directory")
+    };
```

Comment:
I think it would be easier to read if you chain `else` branch with `unwrap_or_else()` too. Also add a comment on why you call `get_relative_target()` twice, very non-intuitive for those who didn't look into how `WasmBuilder` works.